core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/ForwardChainingSchemaCachingRDFSInferencerConnection.java

```diff
@@ -0,0 +1,1126 @@
+            .keySet()
+            .stream()
+            .filter(key -> !calculatedRangeOrDomain.containsKey(key))
+            .forEach(key -> calculatedRangeOrDomain.put(key, new HashSet<>()));
+
+        // Fixed point approach to finding all ranges or domains.
+        // prevSize is the size of the previous application of the function
+        // newSize is the size of the current application of the function
+        // Fixed point is reached when they are the same.
+        long prevSize = 0;
+        final long[] newSize = {-1};
+        while (prevSize != newSize[0]) {
+
+            prevSize = newSize[0];
+
+            newSize[0] = 0;
+
+            calculatedRangeOrDomain.forEach((key, value) -> {
+                List<Resource> resolvedBySubProperty = new ArrayList<>();
+                resolveProperties(key).forEach(newPredicate -> {
+                    Set<Resource> iris = calculatedRangeOrDomain.get(newPredicate);
+                    if (iris != null) {
+                        resolvedBySubProperty.addAll(iris);
+                    }
+
+                });
+
+                List<Resource> resolvedBySubClass = new ArrayList<>();
+                value.addAll(resolvedBySubProperty);
+
+
+                value.stream().map(this::resolveTypes).forEach(resolvedBySubClass::addAll);
+
+                value.addAll(resolvedBySubClass);
+
+                newSize[0] += value.size();
+            });
+
+
+        }
+    }
+
+
+    private boolean inferredCleared = false;
+
+    @Override
+    public void clearInferred(Resource... contexts) throws SailException {
+        super.clearInferred(contexts);
+        inferredCleared = true;
+    }
+
+    private long origianlTboxCount = -1;
+
+
+    @Override
+    public void begin() throws SailException {
+        super.begin();
+
+
+    }
+
+    @Override
+    public void begin(IsolationLevel level) throws UnknownSailTransactionStateException {
+        super.begin(level);
+        forwardChainingSchemaCachingRDFSInferencer.readLock(this);
+
+        origianlTboxCount = getTboxCount();
+
+    }
+
+    @Override
+    public void commit() throws SailException {
+        super.commit();
+        forwardChainingSchemaCachingRDFSInferencer.releaseLock(this);
+    }
+
+    private long getTboxCount() {
+        return forwardChainingSchemaCachingRDFSInferencer.subClassOfStatements.size()
+        + forwardChainingSchemaCachingRDFSInferencer.properties.size()
+        + forwardChainingSchemaCachingRDFSInferencer.subPropertyOfStatements.size()
+        + forwardChainingSchemaCachingRDFSInferencer.rangeStatements.size()
+        + forwardChainingSchemaCachingRDFSInferencer.domainStatements.size()
+        + forwardChainingSchemaCachingRDFSInferencer.types.size();
+    }
+
+    @Override
+    protected void doInferencing() throws SailException {
+        prepareIteration();
+
+        if (forwardChainingSchemaCachingRDFSInferencer.schema == null && origianlTboxCount != getTboxCount()) {
```

Comment:
Relying on the count seems like it would fail too often. Ie, if they removed some and then adding the same number as they removed would this fail?